src/Open-SAMP-API/Client/SAMPFunctions.cpp

```diff
@@ -3,57 +3,53 @@
 #include <Shared/PipeMessages.hpp>
 #include <boost/algorithm/string.hpp>
 
-EXPORT int Client::SAMPFunctions::GetServerIP(char *&ip, int max_len)
-{
+int Client::SAMPFunctions::ReadGTACmdArgument(char *option, char *&str, int max_len) {
 	char *szCommandLine = new char[512];
 	ZeroMemory(szCommandLine, 512);
 
-	GTAFunctions::GetGTACommandLine(szCommandLine, 512);
+	if (GTAFunctions::GetGTACommandLine(szCommandLine, 512) == 0)
+		return 0;
 
+	// The gta_sa.exe process of SAMP looks something like this:
+	// PATH/TO/gta_sa.exe -n NAME -h IP -p PORT
+	// In here, we attempt to read any attributes that were passed in there
 	char *context = NULL;
 	char *token = strtok_s(szCommandLine, " ", &context);
 
 	while (token != NULL) {
 		token = strtok_s(NULL, " ", &context);
-		if (strcmp(token, "-h") == 0)
+		if (strcmp(token, option) == 0)
 			break;
 	}
 
 	if (token != NULL)
 	{
 		token = strtok_s(NULL, " ", &context);
-		strcpy_s(ip, max_len, token);
+		strcpy_s(str, max_len, token);
 	}
 
 	delete[] szCommandLine;
 	return token != NULL;
 }
 
-EXPORT int Client::SAMPFunctions::GetServerPort()
+EXPORT int Client::SAMPFunctions::GetServerIP(char *&ip, int max_len)
 {
-	int port = -1;
-	char *szCommandLine = new char[512];
-	ZeroMemory(szCommandLine, 512);
-
-	GTAFunctions::GetGTACommandLine(szCommandLine, 512);
-
-	char *context = NULL;
-	char *token = strtok_s(szCommandLine, " ", &context);
-
-	while (token != NULL) {
-		token = strtok_s(NULL, " ", &context);
-		if (strcmp(token, "-p") == 0)
-			break;
-	}
+	return ReadGTACmdArgument("-h", ip, max_len);
+}
 
-	if (token != NULL)
-	{
-		token = strtok_s(NULL, " ", &context);
-		port = atoi(token);
+EXPORT int Client::SAMPFunctions::GetServerPort()
+{
+	char *portStr = new char[5];
+	ZeroMemory(portStr, 5);
+	if (ReadGTACmdArgument("-p", portStr, 5)) {
```

Comment:
Notice that the maximum allowed port is 65535, so a `max_len` of 5 should suffice.

Reply:
Just add an additional byte, since the string "should" be zero terminated. Furthermore, clean up the ```portStr``` string in all exit paths. You could use the scope exit library of boost. (https://www.boost.org/doc/libs/1_68_0/libs/scope_exit/doc/html/BOOST_SCOPE_EXIT_ALL.html)